### PR TITLE
Framework: Deprecate use of component-file-picker

### PR DESCRIPTION
**Pull request by furf**
Per #5418, incorporate functionality of third-party `component-file-picker` into A8c `file-picker`.

I removed the unused ability to pass `accept` configuration as `String` or `Array` and the unused additional arguments `event` and `input` passed to the `this.props.onPick` callback from the third-party module. This is safer and saner.

The `getFileAttributes` method sets up an attribute configuration object to be splatted onto the `<input type="file">` element. Since React whitelists safe attributes and the `directory` attribute and its prefixed variants are not yet allowed, a fallback method `setFileDirectoryAttribute` is called on `componentDidMount` and `componentDidUpdate` to manually set these attributes on the element. Once the white list is updated to include the `directory` attribute, `setFileDirectoryAttribute`, `componentDidMount` and `componentDidUpdate` can safely be removed. No changes will be necessary to `getFileAttributes`.

/cc @TooTallNate @nb

### CODE REVIEW

File: client/components/file-picker/index.jsx [hunk 1]
```diff
@@ -4,27 +4,66 @@
...
+
+	// This method is only required until React adds directory to their whitelist.
+	// @see https://github.com/facebook/react/issues/3468
+	// Once it is added, `setFileDirectoryAttribute`, `componentDidMount` and
+	// `componentDidUpdate` methods can be removed.
+	// Note: they are already included in the `getFileAttributes` method.
+	setFileDirectoryAttribute() {
+		if ( this.props.directory ) {
+			const fileInput = this.refs.fileInput;
+			fileInput.setAttribute( `webkitdirectory`, '' );
```
Comment (TooTallNate) [inline client/components/file-picker/index.jsx]: Why use ES6 strings for the name here, if not interpolating any values?

**Comment by TooTallNate** [comment]
General comment: The `component-file-picker` module has an optimization that it only injects _one_ `<input>` element into the DOM, and reuses it for each `pick()` call (since only 1 can be active at a time anyways). It appears that this approach renders one input element per `<FilePicker>`.

Not sure how big of a deal that is in practice (we only render one at the moment so it's virtually the same thing), but I could imagine a page that has _n_ number of pickers and we'd be using more memory in that case.

File: client/components/file-picker/index.jsx (hunk 1 above)
Comment (furf) [inline client/components/file-picker/index.jsx]: Good catch. When I initially implemented, I used a _cute_ one-liner to iterate prefixes, like so:

```
'webkit moz ms o '.split(' ').forEach(prefix => fileInput.setAttribute( `${prefix}directory`, '' ));
```

I later decided there was no need for cute overhead (not to be confused with [cuteoverload](http://cuteoverload.com/)) and flattened out the code. 

The remnant ticks are an oversight. Will replace.

**Comment by furf** [comment]
@TooTallNate that's something to consider. If they were any other DOM element, the overhead might be small enough to ignore. Since `<input type="file">` may contain pointers to the file system once files are selected, I am unaware of the memory impact that might incur. (Still, probably low as they would likely be pointers—I can't imagine they would open file contents into memory until a parent `<form>` submit action was called.) 

We could... 

1. Reimplement the Singleton pattern by sharing a private instance of an `<input type="file">`, updating its attributes per React component instance and call reset on render.
2. Or... put a flag around the render of the input—which would lazily render the input and remove it when the file system dialog closes. No more than one input would ever be on page at any given time, but this would remove need for overhead of updating attributes and resetting a single instance.
3. Or... wait until proof of an actual memory issue before optimizing?

**Comment by TooTallNate** [comment]
Simply moving the file input 1000px up doesn't seem to work out so well:

screen shot 2016-06-01 at 3 08 00 pm

Note that IIRC we can't use `display: none` for browser compat issues (i.e. the input dialog wouldn't appear). I don't remember the specifics of that though.

**Comment by furf** [comment]
Ooph. OK. I copied the CSS from the third-party plugin. I will check to see what I missed. _Probably because they were appending to body and not the button?_

It was rendering fine in dev docs but I should look for more places to test. Where can I find Drop Zone?

The fix is probably as simple as making the containing form `width: 0px; height: 0px; overflow: hidden;`.

**Comment by furf** [comment]
@TooTallNate Thanks for the catch on that CSS. I didn't notice the file input sitting outside of the viewport. Predicted fix worked and has been pushed.

**Comment by lancewillett** [comment]
Closing as older and possibly abandoned. Feel free to re-open if still valid and in-progress.

**Pull request closed without merging**